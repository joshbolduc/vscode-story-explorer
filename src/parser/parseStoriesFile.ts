import { tryParseCsf } from './csf/csf';
import { tryParseMdx } from './mdx/mdx';

const csfParser = { parser: tryParseCsf, type: 'csf' } as const;
const mdxParser = { parser: tryParseMdx, type: 'mdx' } as const;

const getParsers = (filePath?: string) => {
  if (filePath?.endsWith('.mdx') || filePath?.endsWith('.md')) {
    return [mdxParser];
  }

  return [csfParser];
};

export const parseStoriesFile = (contents: string, filePath?: string) => {
  const parsers = getParsers(filePath);

  for (const { parser, type } of parsers) {
    const parsed = parser(contents);
    if (
      parsed &&
      // .mdx files don't require a Meta tag or stories, but .stories.mdx files
      // need at least one
      (type !== 'mdx' ||
        parsed.meta.location !== undefined ||
        parsed.stories.length > 0 ||
        !filePath?.endsWith('.stories.mdx'))
    ) {
      return { ...parsed, type };
    }
  }

  return undefined;
};
