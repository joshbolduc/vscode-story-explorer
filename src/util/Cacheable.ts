/**
 * An object that lazily performs a predefined operation and caches its result.
 */
export class Cacheable<T> {
  private cachedResult?: { result: T } | undefined;

  /**
   * Initialize a new Cacheable with a given callback.
   *
   * @param callback A callback that will be lazily invoked when the result is
   * requested, and whose return value will be cached and returned for
   * subsequent requests.
   */
  public constructor(private callback: () => T) {}

  /**
   * Get the result of the operation, invoking it if necessary. If the operation
   * has already been invoked, its result is returned.
   *
   * @returns The result of the operation, cached or otherwise.
   */
  public getResult() {
    if (!this.cachedResult) {
      this.cachedResult = { result: this.callback() };
    }

    return this.cachedResult.result;
  }

  /**
   * Get the cached result of the operation, if any, or a default value if the
   * result has not already been determined.
   */
  public getCachedResult(): T | undefined;
  public getCachedResult<U>(defaultValue: U): T | U;
  public getCachedResult<U>(defaultValue?: U) {
    return this.cachedResult ? this.cachedResult.result : defaultValue;
  }

  /**
   * Replace the callback used to computed the cached result. The cached result,
   * if any, is cleared.
   *
   * @param callback A callback that will be lazily invoked when the result is
   * requested, and whose return value will be cached and returned for
   * subsequent requests.
   */
  public setCallback(callback: () => T) {
    this.clear();
    this.callback = callback;
  }

  /**
   * Clears the cached result. The next time a result is requested, the
   * operation will be reinvoked and its return value will be cached again.
   */
  public clear() {
    this.cachedResult = undefined;
  }
}
