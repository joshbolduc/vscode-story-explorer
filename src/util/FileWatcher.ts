import type { Disposable, FileSystemWatcher, GlobPattern, Uri } from 'vscode';
import { workspace } from 'vscode';

export type WatchListener = (
  uri: Uri,
  type: 'create' | 'change' | 'delete',
) => void;

export class FileWatcher {
  private fsWatcher?: FileSystemWatcher | undefined;

  private readonly glob: GlobPattern;

  private readonly listenerDisposables: Disposable[] = [];

  public constructor(
    patternOrUri: GlobPattern | Uri,
    callback: WatchListener,
    ignoreCreate = false,
    ignoreChange = false,
    ignoreDelete = false,
  ) {
    this.glob =
      typeof patternOrUri === 'string' || 'pattern' in patternOrUri
        ? patternOrUri
        : patternOrUri.fsPath;
    this.init(callback, ignoreCreate, ignoreChange, ignoreDelete);
  }

  public dispose() {
    this.listenerDisposables.forEach((disposable) => {
      disposable.dispose();
    });
    this.fsWatcher?.dispose();
    this.fsWatcher = undefined;
  }

  private init(
    callback: WatchListener,
    ignoreCreate: boolean,
    ignoreChange: boolean,
    ignoreDelete: boolean,
  ) {
    this.fsWatcher = workspace.createFileSystemWatcher(
      this.glob,
      ignoreCreate,
      ignoreChange,
      ignoreDelete,
    );

    this.listenerDisposables.push(
      this.fsWatcher.onDidChange((uri) => {
        callback(uri, 'change');
      }),
      this.fsWatcher.onDidCreate((uri) => {
        callback(uri, 'create');
      }),
      this.fsWatcher.onDidDelete((uri) => {
        callback(uri, 'delete');
      }),
    );
  }
}
