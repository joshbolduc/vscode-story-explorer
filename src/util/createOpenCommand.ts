import type { TextDocumentShowOptions, Uri } from 'vscode';

interface OpenCommandParams {
  title?: string;
  tooltip?: string;
  arguments: readonly [uri: Uri, options?: TextDocumentShowOptions | undefined];
}

export interface OpenCommandStaticProperties {
  command: 'vscode.open';
}

export const createOpenInEditorCommand = (
  uri: Uri,
  options?: TextDocumentShowOptions,
) => {
  return createOpenCommand({
    title: 'Open',
    tooltip: 'Open in Editor',
    arguments: [uri, options],
  });
};

export const createOpenCommand = <T extends OpenCommandParams>({
  title,
  tooltip,
  arguments: args,
}: T) => {
  return {
    command: 'vscode.open',
    title,
    tooltip,
    arguments: args,
  } as T & OpenCommandStaticProperties;
};
