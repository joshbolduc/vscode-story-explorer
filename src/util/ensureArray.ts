export const ensureArray = <T>(valueOrArray: T | T[]): T[] => {
  if (Array.isArray(valueOrArray)) {
    return valueOrArray;
  }

  return [valueOrArray];
};
