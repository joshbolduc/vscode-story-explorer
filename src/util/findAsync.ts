export const findAsync = async <T>(
  arr: readonly T[],
  predicate: (value: T, index: number, obj: readonly T[]) => Promise<boolean>,
) => {
  for (let i = 0; i < arr.length; i++) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    if (await predicate(arr[i]!, i, arr)) {
      return arr[i];
    }
  }
};
