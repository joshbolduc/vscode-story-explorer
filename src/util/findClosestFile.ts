import { Uri } from 'vscode';
import { findFirstExistingFile } from './findFirstExistingFile';
import { getUriAndParentUris } from './getUriAndParentUris';

export const findClosestFile = (
  filenames: readonly string[],
  startingDir: Uri,
) =>
  findFirstExistingFile(
    getUriAndParentUris(startingDir).flatMap((dir) =>
      filenames.map((filename) => Uri.joinPath(dir, filename)),
    ),
  );
