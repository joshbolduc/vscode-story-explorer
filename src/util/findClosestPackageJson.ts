import type { Uri } from 'vscode';
import { findClosestFile } from './findClosestFile';

export const findClosestPackageJson = (startingDir: Uri) =>
  findClosestFile(['package.json'], startingDir);
