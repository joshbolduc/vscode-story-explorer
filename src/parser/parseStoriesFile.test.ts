import { readFileSync } from 'fs';
import { basename, join } from 'path';
import { sync } from 'fast-glob';
import { describe, expect, it } from 'vitest';
import { testBaseDir } from '../../test/util/testBaseDir';
import { parseStoriesFile } from './parseStoriesFile';

describe('parseStoriesFile', () => {
  const tests = sync(
    [
      'project/v6/src/autoTitle/**/*.stories.*',
      'project/v6/src/stories/v6-only/**/*.stories.*',
      'project/v7/src/**/*.stories.*',
      'project/v7/src/**/*.mdx',
      'fixtures/**/*',
    ],
    {
      cwd: testBaseDir,
    },
  );

  tests.forEach((match) => {
    it(`parses story file ${match}`, () => {
      const storiesPath = join(testBaseDir, match);
      const contents = readFileSync(storiesPath).toString();
      const result = parseStoriesFile(contents, basename(match));

      expect(result).toMatchSnapshot();
    });
  });
});
