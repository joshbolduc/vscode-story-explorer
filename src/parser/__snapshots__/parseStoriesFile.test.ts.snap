// Vitest Snapshot v1

exports[`parseStoriesFile > parses story file fixtures/Children.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": undefined,
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 1, Col 0],
      "name": "Children prop, children content",
      "nameForId": undefined,
    },
    {
      "location": [Ln 8, Col 0],
      "name": "Children prop, empty body",
      "nameForId": undefined,
    },
    {
      "location": [Ln 10, Col 0],
      "name": "Children prop, no body",
      "nameForId": undefined,
    },
    {
      "location": [Ln 12, Col 0],
      "name": "No children prop, children content",
      "nameForId": undefined,
    },
  ],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file fixtures/Mdx1WithStories.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 6, Col 0],
    "title": "Example/MDX with Stories",
  },
  "stories": [
    {
      "location": [Ln 10, Col 0],
      "name": "MDX first story",
      "nameForId": undefined,
    },
    {
      "location": [Ln 14, Col 0],
      "name": "MDX second story",
      "nameForId": undefined,
    },
    {
      "location": [Ln 18, Col 0],
      "name": "MDX Story 3",
      "nameForId": undefined,
    },
  ],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file fixtures/Spread.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": undefined,
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 15, Col 0],
      "name": undefined,
      "nameForId": undefined,
    },
  ],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/autoTitle/csf/EmptyDefaultExport.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 5, Col 17],
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 9, Col 13] - [Ln 9, Col 20],
      "name": "Primary",
      "nameForId": "Primary",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/autoTitle/csf/EmptyTitle.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "title": "",
  },
  "stories": [
    {
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "name": "Primary",
      "nameForId": "Primary",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/autoTitle/csf/NoDefaultExport.stories.js 1`] = `undefined`;

exports[`parseStoriesFile > parses story file project/v6/src/autoTitle/mdx/MdxMeta.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 3, Col 0],
    "title": undefined,
  },
  "stories": [],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/autoTitle/mdx/MdxMetaAndStory.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 3, Col 0],
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 5, Col 0],
      "name": "Primary",
      "nameForId": undefined,
    },
  ],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/autoTitle/mdx/MdxStory.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": undefined,
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 3, Col 0],
      "name": "Primary",
      "nameForId": undefined,
    },
  ],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/autoTitle/mdx/PlainMarkdown.stories.mdx 1`] = `undefined`;

exports[`parseStoriesFile > parses story file project/v6/src/stories/v6-only/TitleInterpolation.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 7, Col 15] - [Ln 10, Col 1],
    "title": "Example/Title Interpolation",
  },
  "stories": [
    {
      "location": [Ln 14, Col 13] - [Ln 14, Col 20],
      "name": "Primary",
      "nameForId": "Primary",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/stories/v6-only/TopHoisted.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 8, Col 1],
    "title": "Top Hoisted",
  },
  "stories": [
    {
      "location": [Ln 12, Col 13] - [Ln 12, Col 23],
      "name": "Top Hoisted",
      "nameForId": "Top Hoisted",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/stories/v6-only/TypeScriptIdentifierDefaultAsMeta.stories.ts 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 6, Col 6] - [Ln 8, Col 1],
    "title": "Test/TypeScript/Identifier Default Exported as Meta",
  },
  "stories": [
    {
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "name": "Type Script Story",
      "nameForId": "Type Script Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/stories/v6-only/TypeScriptIdentifierDefaultSatisfiesAsMeta.stories.ts 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 6, Col 6] - [Ln 8, Col 1],
    "title": "Test/TypeScript/Identifier exported satisfies as Meta",
  },
  "stories": [
    {
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "name": "Type Script Story",
      "nameForId": "Type Script Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/stories/v6-only/TypeScriptIdentifierDefaultSatisfiesMeta.stories.ts 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 6, Col 6] - [Ln 8, Col 1],
    "title": "Test/TypeScript/Identifier exported satisfies Meta",
  },
  "stories": [
    {
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "name": "Type Script Story",
      "nameForId": "Type Script Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v6/src/stories/v6-only/TypeScriptIdentifierSatisfiesAsMeta.stories.ts 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 6, Col 6] - [Ln 8, Col 24],
    "title": "Test/TypeScript/Identifier Satisfies as Meta",
  },
  "stories": [
    {
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "name": "Type Script Story",
      "nameForId": "Type Script Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/autoTitle/csf/ExtraTitle.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "title": "Explicit title",
  },
  "stories": [
    {
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "name": "Primary",
      "nameForId": "Primary",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/autoTitle/csf/nested/WithoutTitle.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "name": "Primary",
      "nameForId": "Primary",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/autoTitle/csf/nested/nested.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "name": "Primary",
      "nameForId": "Primary",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/autoTitle/csf/withIndex/index.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 11, Col 13] - [Ln 11, Col 20],
      "name": "Primary",
      "nameForId": "Primary",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/BasicCsf.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 11, Col 1],
    "title": "Example/Basic CSF",
  },
  "stories": [
    {
      "location": [Ln 15, Col 13] - [Ln 15, Col 20],
      "name": "Primary",
      "nameForId": "Primary",
    },
    {
      "location": [Ln 20, Col 13] - [Ln 20, Col 22],
      "name": "Secondary",
      "nameForId": "Secondary",
    },
    {
      "location": [Ln 25, Col 13] - [Ln 25, Col 18],
      "name": "Large",
      "nameForId": "Large",
    },
    {
      "location": [Ln 31, Col 13] - [Ln 31, Col 18],
      "name": "Small",
      "nameForId": "Small",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/CsfV1.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 1, Col 15] - [Ln 3, Col 1],
    "title": "Example/CSF v1",
  },
  "stories": [
    {
      "location": [Ln 5, Col 13] - [Ln 5, Col 24],
      "name": "Renamed via non-hoisted story",
      "nameForId": "V 1 Story Name",
    },
    {
      "location": [Ln 11, Col 13] - [Ln 11, Col 26],
      "name": "v2 name override",
      "nameForId": "V 2 V 1 Story Name",
    },
    {
      "location": [Ln 18, Col 13] - [Ln 18, Col 31],
      "name": "v1 name should be used",
      "nameForId": "V 2 V 1 Story Name Empty",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/CsfWithMdx.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 0],
    "title": "Example/MDX importing CSF",
  },
  "stories": [
    {
      "location": [Ln 11, Col 0],
      "name": "Basic",
      "nameForId": "Basic",
    },
    {
      "location": [Ln 21, Col 0],
      "name": "Story name from MDX",
      "nameForId": "Basic 2",
    },
    {
      "location": [Ln 25, Col 0],
      "name": "Basic 3",
      "nameForId": "Basic 3",
    },
  ],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/CustomId.stories.js 1`] = `
{
  "meta": {
    "id": "manually-specified-csf-story-id",
    "location": [Ln 1, Col 15] - [Ln 4, Col 1],
    "title": "Example/Custom ID CSF",
  },
  "stories": [
    {
      "location": [Ln 6, Col 13] - [Ln 6, Col 18],
      "name": "Story",
      "nameForId": "Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/CustomId.stories.mdx 1`] = `
{
  "meta": {
    "id": "manually-specified-mdx-id",
    "location": [Ln 3, Col 0],
    "title": "Example/Custom ID MDX",
  },
  "stories": [
    {
      "location": [Ln 7, Col 0],
      "name": "Story",
      "nameForId": undefined,
    },
  ],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/CustomName.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 12, Col 1],
    "title": "Example/Custom Name",
  },
  "stories": [
    {
      "location": [Ln 16, Col 13] - [Ln 16, Col 23],
      "name": "Custom story name",
      "nameForId": "Custom Name",
    },
    {
      "location": [Ln 22, Col 13] - [Ln 22, Col 23],
      "name": "Custom story name override",
      "nameForId": "Overridden",
    },
    {
      "location": [Ln 29, Col 13] - [Ln 29, Col 29],
      "name": "Custom story name var",
      "nameForId": "Named Via Variable",
    },
    {
      "location": [Ln 36, Col 13] - [Ln 36, Col 29],
      "name": "Custom story name var in template string",
      "nameForId": "Interpolated Name",
    },
    {
      "location": [Ln 39, Col 13] - [Ln 39, Col 22],
      "name": "Empty Name",
      "nameForId": "Empty Name",
    },
    {
      "location": [Ln 42, Col 13] - [Ln 42, Col 30],
      "name": "     Leading whitespace",
      "nameForId": "Leading Whitespace",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/DeepTree.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 3, Col 15] - [Ln 6, Col 1],
    "title": "Example/Deep/Tree/With/Several/Kinds",
  },
  "stories": [
    {
      "location": [Ln 10, Col 13] - [Ln 10, Col 18],
      "name": "Story",
      "nameForId": "Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/DocsOnly.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 3, Col 0],
    "title": "Example/MDX Docs Only",
  },
  "stories": [],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/DocsOnlyTop.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 3, Col 0],
    "title": "MDX Docs at top level",
  },
  "stories": [],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/Excluded.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 4, Col 15] - [Ln 7, Col 1],
    "title": "Test/Exclusions/Excluded",
  },
  "stories": [
    {
      "location": [Ln 2, Col 13] - [Ln 2, Col 24],
      "name": "Other Export",
      "nameForId": "Other Export",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/ExportStyles.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 11, Col 1],
    "title": "Example/Export Styles",
  },
  "stories": [
    {
      "location": [Ln 15, Col 6] - [Ln 15, Col 18],
      "name": "Normal Export",
      "nameForId": "Normal Export",
    },
    {
      "location": [Ln 20, Col 6] - [Ln 20, Col 16],
      "name": "Alternative Export Name",
      "nameForId": "Alternative Export Name",
    },
    {
      "location": [Ln 25, Col 6] - [Ln 25, Col 29],
      "name": "Override name 3",
      "nameForId": "Export Name Overridden",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/Form.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 3, Col 15] - [Ln 5, Col 1],
    "title": "Example/Form",
  },
  "stories": [
    {
      "location": [Ln 7, Col 13] - [Ln 7, Col 17],
      "name": "Form",
      "nameForId": "Form",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/Hoisting.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 7, Col 1],
    "title": "Example/Hoisting",
  },
  "stories": [
    {
      "location": [Ln 3, Col 13] - [Ln 3, Col 21],
      "name": "Hoisting",
      "nameForId": "Hoisting",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/Included.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 4, Col 15] - [Ln 7, Col 1],
    "title": "Test/Exclusions/Included",
  },
  "stories": [
    {
      "location": [Ln 1, Col 13] - [Ln 1, Col 27],
      "name": "Included Export",
      "nameForId": "Included Export",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/JsxExtension.stories.jsx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 1, Col 15] - [Ln 3, Col 1],
    "title": "Example/JSX extension",
  },
  "stories": [
    {
      "location": [Ln 5, Col 13] - [Ln 5, Col 18],
      "name": "Story",
      "nameForId": "Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/Mdx2WithStories.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 12, Col 0],
    "title": "Example/MDX with Stories",
  },
  "stories": [
    {
      "location": [Ln 14, Col 0],
      "name": "MDX first story",
      "nameForId": undefined,
    },
    {
      "location": [Ln 18, Col 0],
      "name": "MDX second story",
      "nameForId": undefined,
    },
    {
      "location": [Ln 22, Col 0],
      "name": "MDX Story 3",
      "nameForId": undefined,
    },
  ],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/MdxWithStoriesTop.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 4, Col 0],
    "title": "MDX with stories, top",
  },
  "stories": [
    {
      "location": [Ln 6, Col 0],
      "name": "MDX with stories, top",
      "nameForId": undefined,
    },
  ],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/MergeOne.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 1, Col 15] - [Ln 3, Col 1],
    "title": "Example/Merge",
  },
  "stories": [
    {
      "location": [Ln 5, Col 13] - [Ln 5, Col 16],
      "name": "One",
      "nameForId": "One",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/MergeTwo.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 1, Col 15] - [Ln 3, Col 1],
    "title": "Example/Merge",
  },
  "stories": [
    {
      "location": [Ln 5, Col 13] - [Ln 5, Col 16],
      "name": "Two",
      "nameForId": "Two",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/Precedence.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 4, Col 0],
    "title": "Example/Precedence",
  },
  "stories": [
    {
      "location": [Ln 20, Col 0],
      "name": "Story with name, story, body",
      "nameForId": "Small",
    },
    {
      "location": [Ln 26, Col 0],
      "name": "Story with name, body",
      "nameForId": undefined,
    },
  ],
  "type": "mdx",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/SameName.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 8, Col 1],
    "title": "Example/With Same Name",
  },
  "stories": [
    {
      "location": [Ln 12, Col 13] - [Ln 12, Col 21],
      "name": "Same Name",
      "nameForId": "Same Name",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/SameNameChild.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 8, Col 1],
    "title": "Example/With Same Name/Same Name",
  },
  "stories": [
    {
      "location": [Ln 12, Col 13] - [Ln 12, Col 24],
      "name": "Nested Story",
      "nameForId": "Nested Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/StoryDeclarations.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 1, Col 6] - [Ln 3, Col 1],
    "title": "Test/Story Declarations",
  },
  "stories": [
    {
      "location": [Ln 7, Col 13] - [Ln 7, Col 28],
      "name": "Const Declarator",
      "nameForId": "Const Declarator",
    },
    {
      "location": [Ln 9, Col 6] - [Ln 9, Col 24],
      "name": "Separate Declarator",
      "nameForId": "Separate Declarator",
    },
    {
      "location": [Ln 10, Col 6] - [Ln 10, Col 25],
      "name": "Separate declarator with story name",
      "nameForId": "Separate Declarator 2",
    },
    {
      "location": [Ln 15, Col 13] - [Ln 15, Col 24],
      "name": "Multiple One",
      "nameForId": "Multiple One",
    },
    {
      "location": [Ln 16, Col 2] - [Ln 16, Col 13],
      "name": "Multiple two (given name)",
      "nameForId": "Multiple Two",
    },
    {
      "location": [Ln 20, Col 16] - [Ln 20, Col 29],
      "name": "Renamed Function Story",
      "nameForId": "Function Story",
    },
    {
      "location": [Ln 25, Col 9] - [Ln 25, Col 30],
      "name": "Separate Function Story",
      "nameForId": "Separate Function Story",
    },
    {
      "location": [Ln 29, Col 9] - [Ln 29, Col 31],
      "name": "Separate function story with story name",
      "nameForId": "Separate Function Story 2",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/Top.stories.js 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 8, Col 1],
    "title": "Top",
  },
  "stories": [
    {
      "location": [Ln 12, Col 13] - [Ln 12, Col 26],
      "name": "Top Level Story",
      "nameForId": "Top Level Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/TypeScript.stories.tsx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 17, Col 15] - [Ln 19, Col 9],
    "title": "Test/TypeScript/Sample",
  },
  "stories": [
    {
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "name": "Type Script Story",
      "nameForId": "Type Script Story",
    },
    {
      "location": [Ln 5, Col 13] - [Ln 5, Col 40],
      "name": "Type Script Typed Story",
      "nameForId": "Type Script Typed Story",
    },
    {
      "location": [Ln 9, Col 13] - [Ln 9, Col 51],
      "name": "Type Script Typed Generic Story",
      "nameForId": "Type Script Typed Generic Story",
    },
    {
      "location": [Ln 13, Col 13] - [Ln 13, Col 41],
      "name": "Type Script Type Assertion Story",
      "nameForId": "Type Script Type Assertion Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/TypeScriptAsMeta.stories.ts 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 15] - [Ln 7, Col 9],
    "title": "Test/TypeScript/Default as Meta",
  },
  "stories": [
    {
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "name": "Type Script Story",
      "nameForId": "Type Script Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/TypeScriptIdentifier.stories.ts 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 6] - [Ln 7, Col 1],
    "title": "Test/TypeScript/Typed Identifier",
  },
  "stories": [
    {
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "name": "Type Script Story",
      "nameForId": "Type Script Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/TypeScriptIdentifierAsMeta.stories.ts 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 6, Col 6] - [Ln 8, Col 9],
    "title": "Test/TypeScript/Identifier as Meta",
  },
  "stories": [
    {
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "name": "Type Script Story",
      "nameForId": "Type Script Story",
    },
  ],
  "type": "csf",
}
`;

exports[`parseStoriesFile > parses story file project/v7/src/stories/TypeScriptIdentifierSatisfies.stories.ts 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 6, Col 6] - [Ln 8, Col 16],
    "title": "Test/TypeScript/Identifier Satisfies",
  },
  "stories": [
    {
      "location": [Ln 3, Col 13] - [Ln 3, Col 28],
      "name": "Type Script Story",
      "nameForId": "Type Script Story",
    },
  ],
  "type": "csf",
}
`;
